### PR TITLE
RIIR: The Inevitable Rewrite It In Rust

### PR DESCRIPTION
**Pull request by ErinvanderVeen**
This PR completely reimplements the nixtract logic in Rust.

For this description, I will refer to the current python implementation as `nixtract`, and the rust rewrite as `nixtract-rs`.

The goal of this rewrite is multifaceted:

1. Performance. `nixtract-rs` does not have a single locking shared job queue, instead giving each worker its own queue thanks to the `rayon` crate. This ensures we can spend up to 99% of the total program run time performing Nix evaluation. For reference, extracting `firefox` from current nixpkgs master:
```fish
❯ time nixtract --n-workers 12 --target-flake-ref "github:NixOS/nixpkgs?rev=1828ca4dc62c453c9132258d37775eb28ca43e25" --target-attribute-path firefox - > /dev/null
________________________________________________________
Executed in   54.49 secs    fish           external
   usr time  315.64 secs    0.00 micros  315.64 secs
   sys time   74.31 secs  876.00 micros   74.31 secs


❯ time nixtract-rs --target-flake-ref "github:NixOS/nixpkgs?rev=1828ca4dc62c453c9132258d37775eb28ca43e25" --target-attribute-path firefox > /dev/null
________________________________________________________
Executed in   50.81 secs    fish           external
   usr time  314.60 secs  107.00 micros  314.60 secs
   sys time   73.75 secs  261.00 micros   73.75 secs

```
2. Maintainability. `nixtract` is written in Python, which I @ErinvanderVeen have very limited experience with.
3. The main usecase of `nixtract` is its use in `Genealogos`, a Rust project. `nixtract-rs` can be linked statically with `genealogos`, which makes running the tool easier.

### CODE REVIEW

File: src/nix/describe_derivation.nix [hunk 1]
```diff
@@ -10,12 +10,12 @@
 
 let
   nixpkgs = builtins.getFlake "nixpkgs";
-  lib = import ./lib.nix { inherit nixpkgs; };
+  lib = import <lib> { inherit nixpkgs; };
```
Comment (GuillaumeDesforges) [inline src/nix/describe_derivation.nix:13]: Why did you need to change this?

File: src/nix/find_attribute_paths.nix [hunk 2]
```diff
@@ -13,12 +13,12 @@
 
 let
   nixpkgs = builtins.getFlake "nixpkgs";
-  lib = import ./lib.nix { inherit nixpkgs; };
+  lib = import <lib> { inherit nixpkgs; };
```
Comment (GuillaumeDesforges) [inline src/nix/find_attribute_paths.nix:16]: Same question

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                &collected_paths,
+                flake_ref,
+                system,
+                &found_drv.attribute_path,
+                offline,
+                &lib,
+            )
+        })
+        .flatten()
+        .collect();
```
Comment (GuillaumeDesforges) [inline src/lib.rs:114]: It looks to me like this implementation can describe many times the same derivation, whereas the current (Python) one does not (using a mutable set to check if a derivation has already been described).

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            .unwrap_or("builtins.currentSystem"),
+        attribute_path.map(AsRef::as_ref).unwrap_or("")
+    );
+
+    let collected_paths: Arc<Mutex<std::collections::HashSet<String>>> =
+        Arc::new(Mutex::new(std::collections::HashSet::new()));
+
+    // call find_attribute_paths to get the initial set of derivations
+    let attribute_paths =
+        nix::find_attribute_paths(flake_ref, system, attribute_path, offline, &lib)?;
```
Comment (GuillaumeDesforges) [inline src/lib.rs:92]: Would it be possible, like the current (Python) implementation, to start describing derivations before this eval finishes (as it can be pretty long)?

File: src/nix/describe_derivation.nix (hunk 1 above)
Comment (ErinvanderVeen) [reply src/nix/describe_derivation.nix:13]: Distributing via crates.io doesn't currently allow resource files to be included with your program. As such we need to write lib.nix to disk somewhere (which we do, to a /tmp/ file). This file is then added to the NIX_PATH, hence we need `<lib>`.

File: src/lib.rs (hunk 3 above)
Comment (ErinvanderVeen) [reply src/lib.rs:114]: I don't think it does: https://github.com/tweag/nixtract/pull/26/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R29

File: src/lib.rs (hunk 4 above)
Comment (ErinvanderVeen) [reply src/lib.rs:92]: Sure, it would be possible. But this current approach is nice and simple, and `nixtract-rs` will mainly be used as a library anyway.

File: src/nix/describe_derivation.nix (hunk 1 above)
Comment (ErinvanderVeen) [reply src/nix/describe_derivation.nix:13]: We could technically pass the path to the lib.nix file in another env-var, but NIX_PATH already exists, so we might as well make use  of it.

**Review by dorranh — COMMENTED** [review]
Nice work, @ErinvanderVeen! One question would be whether we've done a 1:1 comparison of the output of the Python and Rust implementations for a given Nix derivation just as a sanity check?

**Comment by ErinvanderVeen** [comment]
I have done a 1:1 comparison indeed. This just highlighted a bug in the python implementation where it can sometimes happen that it outputs the same component twice. Other than that, the output is identical.
